contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,594 @@
+is to allow administrators to request fail open on specific calls, or to require that certain special resource
+paths (initializers dynamic config path) are always fail open. Alternatively, it may be desirable for
+administrative users to bypass admission completely.
+
+Some options:
+
+1. Some namespaces are opted out of external admission (kube-system)
+2. Certain service accounts can bypass external admission checks
+
+
+##### Upgrade of a cluster with external admission
+
+The current order of cluster upgrade is `apiserver` -> `controller` -> `nodes`.  External admission controllers
+would typically need to be upgraded first in order to ensure new semantic changes in objects are not ignored.
+This would include fields like PodSecurityContext - adding that *prior* to admission is necessary because it
+allows escalation that was previously impossible.
+
+
+#### Dynamic configuration
+
+Trusted clients should be able to modify the initialization and external admission hooks on the fly and expect
+that configuration is updated quickly. Extension API servers should also be able to leverage central
+configuration, but may opt for alternate mechanisms.
+
+The initializer admission controller and the generic webhook admission controller should dynamically load config
+from a `ConfigMap` or a net new API object holding the following configuration schema:
+
+```
+type AdmissionControlConfiguration struct {
+    TypeMeta // although this object could simply be serialized like ComponentConfig
+
+    // ResourceInitializers is a list of resources and their default initializers
+    ResourceInitializers []ResourceDefaultInitializer
+
+    ExternalAdmissionHooks []ExternalAdmissionHook
+}
+
+type ResourceDefaultInitializer struct {
+    // Resource identifies the type of resource to be initialized that should be initialized
+    Resource GroupResource
+    // Initializers are the default names that will be registered to this resource
+    Initializers []string
+}
+
+type ExternalAdmissionHook struct {
+    // Operations is the list of operations this hook will be invoked on - Create, Update, or *
+    // for all operations. Defaults to '*'.
+    Operations []string
+    // Resources are the resources this hook should be invoked on. '*' is all resources.
+    Resources []string
+    // Subresources are the list of subresources this hook should be invoked on. '*' is all resources.
+    Subresources []string
+
+    // TODO define client configuration
+
+    // FailurePolicy defines how unrecognized errors from the admission endpoint are handled -
+    // allowed values are Ignore, Retry, Fail. Default value is Fail
+    FailurePolicy FailurePolicyType
+}
+```
+
+All changes to this config must be done in an safe matter - when adding a new hook or initializer,
+first verify the new agent is online before allowing it to come into rotation. Removing a hook or
+initializer must occur before disabling the remote endpoint, and all queued items must complete.
```

Comment:
IMO same measure should be taken if the agent is not online yet or if the agent dies afterwards. In the current design, if the agent dies afterwards, for initializers, we always fail closed; for webhooks, users can choose between fail open and fail closed.

"Checking if the agent is online before adding the initializer/webhook into rotation" equals to "fail open until succeed once". Sometimes it's a nice feature to have, but doesn't help much if the agent dies 10s after an initial success. Also, some initializers/webhooks are so import that users want them to fail closed from the beginning, which is the opposite of the proposed behavior.

So i think the system shouldn't treat adding an agent as a special case.

Replying to some offline discussion:
> By analogy, we don't immediately put a pod into a service rotation until its ready

There is a difference. The endpoints controller will remove the pod from service rotation if the pod becomes "not ready" afterwards. So this system always fails open, no matter if a pod is not ready at the starting time, or fails later.

> Anyway, you have to wait for all apiservers to reach consensus on the set of applied config, so there's some sync anyway.

Are we talking about HA mode? In that case, the `initializer admission controller` and the `generic webhook admission controller` should make a quorum read of the dynamic configuration, so the synchronization is already handled by etcd.